drivers/dma/dma_sam0.c

```diff
@@ -0,0 +1,242 @@
+/*
+ * Copyright (c) 2018 Google LLC.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_DMA_LEVEL
+#define SYS_LOG_DOMAIN "dma"
+#include <logging/sys_log.h>
+
+#include <board.h>
+#include <dma.h>
+
+typedef void (*dma_callback)(struct device *dev, u32_t channel,
+			     int error_code);
+
+struct dma_sam0_device {
+	__aligned(16) DmacDescriptor descriptors[DMAC_CH_NUM];
+	__aligned(16) DmacDescriptor descriptors_wb[DMAC_CH_NUM];
+	dma_callback cb[DMAC_CH_NUM];
+};
+
+/* Handles DMA channel interrupts and dispatches to the callback */
+static void dma_sam0_channel_isr(struct device *dev,
+				 struct dma_sam0_device *data, int channel)
+{
+	u8_t ints = DMAC->CHINTFLAG.reg & DMAC->CHINTENSET.reg;
+	dma_callback cb = data->cb[channel];
+
+	/* Acknowledge all interrupts */
+	DMAC->CHINTFLAG.reg = ints;
+
+	if (cb != NULL) {
+		if ((ints & DMAC_CHINTFLAG_TCMPL) != 0) {
+			cb(dev, channel, 0);
+		}
+		if ((ints & DMAC_CHINTFLAG_TERR) != 0) {
+			cb(dev, channel, DMAC_CHINTFLAG_TERR);
+		}
+	}
+}
+
+/* Handles DMA interrupts and dispatches to the individual channel */
+static void dma_sam0_isr(void *arg)
+{
+	struct device *dev = arg;
+	struct dma_sam0_device *data = dev->driver_data;
+	int channel;
+	u32_t intstatus = DMAC->INTSTATUS.reg;
+
+	for (channel = 0; intstatus != 0; intstatus >>= 1) {
```

Comment:
Given that there's a small possibility of having a lot of interrupts pending when this function is called, it might be better to use find_lsb_set() here:

```C
while (true) {
    int channel = find_lsb_set(intstatus);

    if (channel == 0) {
        break;
    }

    dma_sam0_channel_isr(dev, data, channel);
    intstatus &= ~BIT(channel);
}
```

Reply:
Hi Leandro.  The Cortex-M0+ doesn't have a clz instruction so find_lsb_set becomes a function call.  I'm not sure it'd be faster than this loop, especially as it early terminates.

Reply:
Ah. LGTM then.